### PR TITLE
test creation

### PR DESCRIPTION
**Pull request by Clio18**

### CODE REVIEW

File: src/main/java/dao/DrugDAO.java [hunk 1]
```diff
@@ -4,15 +4,26 @@
...
     }
 
+    public MockDrugBase getMockDrugBase() {
+        return mockDrugBase;
+    }
+
+    public void setMockDrugBase(MockDrugBase mockDrugBase) {
+        this.mockDrugBase = mockDrugBase;
+    }
+
```
Comment (tsupryk) [inline src/main/java/dao/DrugDAO.java:38]: Why do you use get/set for mockDrugBase and constructor with List<Drug> drugs to be setted in mockDrugBase ? 
You will loose consistency if someone will accidentally call setMockDrugBase(...)

File: src/main/java/dao/DrugDAO.java [hunk 2]
```diff
@@ -4,15 +4,26 @@
 import source.MockDrugBase;
 
 import java.util.ArrayList;
+import java.util.Iterator;
 import java.util.List;
 
 public class DrugDAO implements Dao<Drug> {
-    MockDrugBase mockDrugBase = new MockDrugBase();
+   private MockDrugBase mockDrugBase;
```
Comment (tsupryk) [inline src/main/java/dao/DrugDAO.java]: field name with 'mock' ? maybe something like drugsStorage is better ?

File: src/main/java/dao/DrugDAO.java [hunk 3]
```diff
@@ -4,15 +4,26 @@
...
+    public MockDrugBase getMockDrugBase() {
+        return mockDrugBase;
+    }
+
+    public void setMockDrugBase(MockDrugBase mockDrugBase) {
+        this.mockDrugBase = mockDrugBase;
+    }
+
     @Override
     public Drug get(int id) throws CloneNotSupportedException {
```
Comment (tsupryk) [inline src/main/java/dao/DrugDAO.java]: Please rename it to findById(int id) - it's a bit confusing

File: src/test/java/DrugDAOTest.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    @Test
+    public void whenUpdateDrug() throws CloneNotSupportedException {
+        Drug updatedDrug = new Drug("Az", "AAz", 1.1, false, DosageForm.HARD_CAPSULE, "AAAz", 1);
+        drugDAO.update(updatedDrug);
+        assertEquals(updatedDrug, drugDAO.get(1));
+    }
+
+    @Test
+    public void whenFindDrugByName() throws CloneNotSupportedException {
+        assertEquals(drugDAO.findByName("AA"), drugDAO.getAll().get(0));
```
Comment (tsupryk) [inline src/test/java/DrugDAOTest.java]: Please don't rely on elements order in your tests: drugDAO.getAll().get(0). What if you'll decide to switch dao implementation from ArrayList to HashMap ?

File: src/test/java/DrugDAOTest.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    }
+
+    @Test
+    public void whenReturnAll() {
+        assertThat(drugDAO.getAll().size(), is(3));
+    }
+
+    @Test
+    public void whenFindByPrice() {
+        List <Drug> finded = drugDAO.findDrugsByPrice(1.0, 2.0);
```
Comment (tsupryk) [inline src/test/java/DrugDAOTest.java]: not finded but found :)

**Review by tsupryk — CHANGES_REQUESTED** [review]
Tests are great!! 
They can be even better with testing of corner cases: 
- try to find drug with not existing id
- update not exising drug
- delete not existing drug

File: src/main/java/dao/DrugDAO.java [hunk 6]
```diff
@@ -1,22 +1,32 @@
...
 public class DrugDAO implements Dao<Drug> {
-    MockDrugBase mockDrugBase = new MockDrugBase();
+   private MockDrugBase drugsStorage;
 
-    public DrugDAO() {
+    public DrugDAO(List<Drug> drugs) {
+        drugsStorage = new MockDrugBase(drugs);
+    }
+
+    public MockDrugBase getDrugsStorage() {
```
Comment (tsupryk) [inline src/main/java/dao/DrugDAO.java]: Remove this getter

File: src/main/java/dao/DrugDAO.java [hunk 7]
```diff
@@ -27,46 +37,65 @@ public Drug get(int id) throws CloneNotSupportedException {
...
+//                oldDrug.setName(drug.getName());
+//                oldDrug.setPrice(drug.getPrice());
+//                oldDrug.setManufacturer(drug.getManufacturer());
+//                oldDrug.setDosageForm(drug.getDosageForm());
+//                oldDrug.setDescription(drug.getDescription());
+//                oldDrug.setAvailable(drug.isAvailable());
+//            }else {
+//                throw new DrugNotFoundException();
+//            }
+//        }
```
Comment (tsupryk) [inline src/main/java/dao/DrugDAO.java]: remove commented code

File: src/main/java/dao/DrugDAO.java [hunk 8]
```diff
@@ -27,46 +37,65 @@ public Drug get(int id) throws CloneNotSupportedException {
...
+                d.setManufacturer(drug.getManufacturer());
+                d.setDosageForm(drug.getDosageForm());
+                d.setDescription(drug.getDescription());
+                d.setAvailable(drug.isAvailable());
+                drugForUpdate = d;
             }
         }
+        if (drugForUpdate==null){
+            throw new DrugNotFoundException();
+        }
```
Comment (tsupryk) [inline src/main/java/dao/DrugDAO.java]: It's easier to read when code has sequence:
1. find Drug for update
2. check if its not null
3. updated Drug fields

File: src/main/java/exception/DrugNotFoundException.java [hunk 9]
```diff
@@ -0,0 +1,9 @@
+package exception;
+
+public class DrugNotFoundException extends Exception {
```
Comment (tsupryk) [inline src/main/java/exception/DrugNotFoundException.java:3]: why this exception is checked?

File: src/main/java/dao/DrugDAO.java [hunk 10]
```diff
@@ -27,46 +37,65 @@ public Drug get(int id) throws CloneNotSupportedException {
...
+//                oldDrug.setAvailable(drug.isAvailable());
+//            }else {
+//                throw new DrugNotFoundException();
+//            }
+//        }
 
     }
 
     @Override
     public void delete(int id) {
```
Comment (tsupryk) [inline src/main/java/dao/DrugDAO.java]: make method return true, if Drug was deleted, false otherwise

File: src/test/java/DrugDAOTest.java [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+        drugDAO.delete(1);
+        assertThat(drugDAO.getDrugsStorage().getPharmacy().size(), is(2));
+
+    }
+
+    @Test
+    public void whenFindDrugById() throws CloneNotSupportedException {
+        assertEquals(drugDAO.findById(1), drug1);
+        assertEquals(drugDAO.findById(900), null);
+    }
```
Comment (tsupryk) [inline src/test/java/DrugDAOTest.java]: please add one more test:
```
drugDAO.save(new Drug("Vitamin", "Vitamin", 1.1, true, DosageForm.HARD_CAPSULE, "AAA", 1));
assertEquals(drugDAO.findById(1), drug1);
```

File: src/main/java/dao/Dao.java [hunk 12]
```diff
@@ -1,13 +1,15 @@
 package dao;
+import exception.DrugNotFoundException;
+
 import java.util.List;
 
 
 public interface Dao<T> {
-    T get(int id) throws CloneNotSupportedException;
+    T findById(int id) throws CloneNotSupportedException;
```
Comment (tsupryk) [inline src/main/java/dao/Dao.java:8]: Why user should care about handling some CloneNotSupportedException when finding an object?

File: src/test/java/DrugDAOTest.java [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+        Drug updatedDrug = new Drug("Az", "AAz", 1.1, false, DosageForm.HARD_CAPSULE, "AAAz", 1);
+        drugDAO.update(updatedDrug);
+        assertEquals(updatedDrug, drugDAO.findById(1));
+
+        Drug updatedDrug2 = new Drug("Az", "AAz", 1.1, false, DosageForm.HARD_CAPSULE, "AAAz", 100);
+        try {
+            drugDAO.update(updatedDrug2);
+        } catch (Exception e){
+            assertEquals(e.getMessage(), new DrugNotFoundException().getMessage());
+        }
```
Comment (tsupryk) [inline src/test/java/DrugDAOTest.java]: there should be 2 separate tests: one for checking correct update, another to check that exception is thrown

**Review by tsupryk — CHANGES_REQUESTED** [review]
Good! Only few things left to fix :)
